File: refsol/add-and-search-word-data-structure-design.js
export class WordDictionary {
  constructor() {
    this.root = {};
  }

  addWord(word) {
    let node = this.root;
    for (const ch of word) {
      if (!node[ch]) node[ch] = {};
      node = node[ch];
    }
    node["$"] = true;
  }

  search(word) {
    const dfs = (node, i) => {
      if (i === word.length) return node["$"] === true;
      const ch = word[i];
      if (ch === ".") {
        for (const key of Object.keys(node)) {
          if (key !== "$" && dfs(node[key], i + 1)) return true;
        }
        return false;
      }
      if (node[ch]) return dfs(node[ch], i + 1);
      return false;
    };
    return dfs(this.root, 0);
  }
}

File: refsol/add-and-search-word-data-structure-design.py
class WordDictionary:
    def __init__(self):
        self.root = {}

    def addWord(self, word):
        node = self.root
        for ch in word:
            node = node.setdefault(ch, {})
        node["$"] = True

    def search(self, word):
        def dfs(node, i):
            if i == len(word):
                return "$" in node
            ch = word[i]
            if ch == ".":
                return any(
                    key != "$" and dfs(child, i + 1)
                    for key, child in node.items()
                )
            if ch in node:
                return dfs(node[ch], i + 1)
            return False

        return dfs(self.root, 0)
